source/slang/slang-ir-spirv-legalize.cpp

```diff
@@ -1167,6 +1167,116 @@ struct SPIRVLegalizationContext : public SourceEmitterBase
         addUsersToWorkList(newStore);
     }
 
+    void processNonUniformResourceIndex(IRInst* nonUniformResourceIndexInst)
+    {
+        // implement the translation to spirv by waking up the use-def chain
+        // from nonUniformResource inst of an index to an array of buffer or
+        // texture def all the way to the leaf operations. To be precise:
+        // - go through GEP and see if it calls an intrinsic function,
+        //   then decorate the address itself (GetElementPtr)
+        // - go through GEP to identify the pointer access and the Loads that it
+        //   accesses (GetElementPtr -> Load), then decorate the load instruction.
+        // - go through IntCasts to deal with u32 -> i32 / vice-versa (IntCast)
+        List<IRInst*> resWorkList;
+        resWorkList.add(nonUniformResourceIndexInst);
+
+        // iterate down all the insts that originate from a `nonUniformResourceIndexInst` and pop out the
+        // `nonUniformResourceIndexInst` from the index value itself and wrap it around the parent inst.
+        // For example:
+        IRCall* callInst = nullptr;
+        List<IRInst*> newCallArgs;
+        for (Index i = 0; i < resWorkList.getCount(); i++)
+        {
+            auto inst = resWorkList[i];
+            traverseUses(inst, [&](IRUse* use)
+            {
+                auto user = use->getUser();
+                IRBuilder builder(user);
+                builder.setInsertBefore(user);
+                IRInst* newUser = nullptr;
+                switch (user->getOp())
+                {
+                case kIROp_IntCast:
+                    // Replace intCast(nonUniformRes(x)), into nonUniformRes(intCast(x))
+                    newUser = builder.emitCast(user->getFullType(), inst->getOperand(0));
+                    break;
+                case kIROp_GetElementPtr:
+                    // Ignore when nonUniformRes is not on the index
+                    if (user->getOperand(0)->getOp() != kIROp_NonUniformResourceIndex)
+                    {
+                        // Replace gep(pArray, nonUniformRes(x)), into nonUniformRes(gep(pArray, x))
+                        newUser = builder.emitElementAddress(user->getFullType(), user->getOperand(0), inst->getOperand(0));
+                    }
+                    break;
+                case kIROp_RWStructuredBufferGetElementPtr:
+                    // Replace buffer[nonUniformRes(nonUniformRes(x))], into nonUniformRes(buffer[x])
+                    // This fails when converting rwstructbuf(nonuniformres(gep(...), int) -> rwstructbuf(get(...), int).
+                    // newUser = builder.emitRWStructuredBufferGetElementPtr(inst->getOperand(0), user->getOperand(1));
+                    break;
+                case kIROp_Load:
+                    newUser = builder.emitLoad(user->getFullType(), inst->getOperand(0));
+                    break;
+                case kIROp_Call:
```

Comment:
This seems dangerous, why is it needed?

Reply:
It is for the case of 
```
buffer[NonUniformResourceIndex(0)].InterlockedAdd(0, 1);
```
which results in: 
```
kIROp_Call (171) "RWByteAddressBuffer.InterlockedAdd", kIROp_NonUniformResourceIndex, ... )
```
So, I added the above to deal with this. if I ignore it, then the first param `nonUniformResourceIndex(%1)` will result in a compilation failure.

Reply:
But this doesn't seem right? I would say that in this case the IR:

%gep = GetElementAddr ...
%ld = load gep
%buffer = NonUniformIndex ld
IRCall %RWStructuredBuffer_InterlockedAdd buffer ...

Is the right form you want, right?

Reply:
This doesn't seem to be the pattern we get. The rwStructuredBufferGep is on the output (store), the input path is pretty simple:
```
        {674}   let  %32        : Int   = nonUniformResourceIndex(2 : Int)
        {1360}  call %RWByteAddressBufferx5FInterlockedAdd(%32, 0 : UInt, 1 : UInt)
        {1345}  let  %35        : UInt  = call %RWByteAddressBufferx5FLoad(0 : Int, 0 : Int)
        {910}   let  %36        : Float = castIntToFloat(%35)
```
The output path is little more tedious:
```
        {1486}  let  %29        : Vec(UInt, 3 : Int)    = load(%27)
        {1400}  let  %30        : UInt  = swizzle(%29, 0 : Int)
        {1401}  let  %31        : Int   = intCast(%30)
        {863}   let  %33        : UInt  = intCast(%31)
        {882}   let  %34        : Ptr(Float)    = rwstructuredBufferGetElementPtr(%output, %33)
        {911}   store(%34, %36)
```


Sharing the IR that we see in this example for context:
```
{666}   block %28:
        {1486}  let  %29        : Vec(UInt, 3 : Int)    = load(%27)
        {1400}  let  %30        : UInt  = swizzle(%29, 0 : Int)
        {1401}  let  %31        : Int   = intCast(%30)
        {674}   let  %32        : Int   = nonUniformResourceIndex(2 : Int)
        {1360}  call %RWByteAddressBufferx5FInterlockedAdd(%32, 0 : UInt, 1 : UInt)
        {1408}  SPIRVAsm _      : Void
        {
                {1409}  SPIRVAsmInst(225 : UInt, 1 : UInt, 2376 : UInt)
        }

        {863}   let  %33        : UInt  = intCast(%31)
        {882}   let  %34        : Ptr(Float)    = rwstructuredBufferGetElementPtr(%output, %33)
        {1345}  let  %35        : UInt  = call %RWByteAddressBufferx5FLoad(0 : Int, 0 : Int)
        {910}   let  %36        : Float = castIntToFloat(%35)
        {911}   store(%34, %36)
        {913}   return_val(void_constant)
}
```

This is quite different from the other example that you showed where the nonUniform indexing is done within the InterlockedAdd
```
    InterlockedAdd(texArray[NonUniformResourceIndex(dispatchThreadID.x)][uint2(0)], 2);
```

I feel like this is the only way we can deal with this scenario. It is used by the `tests/spirv/coherent-2.slang` example.

Reply:
The change https://github.com/shader-slang/slang/pull/3965 makes this irrelevant issue. The problem of having to decorate a Call to an atomic function will be unnecessary if the call is inlined.